lib/pbio/drv/display/display_ev3.c

```diff
@@ -503,27 +455,67 @@ PROCESS_THREAD(pbdrv_display_ev3_init_process, ev, data) {
     pbdrv_display_load_indexed_bitmap(pbdrv_display_pybricks_logo);
     pbdrv_display_st7586s_encode_user_frame();
     pbdrv_display_st7586s_write_data_begin(st7586s_send_buf, sizeof(st7586s_send_buf));
-    PROCESS_WAIT_EVENT_UNTIL(ev == PROCESS_EVENT_POLL && spi_status == SPI_STATUS_COMPLETE);
+    PBIO_OS_AWAIT_UNTIL(state, spi_status == SPI_STATUS_COMPLETE);
     pbdrv_gpio_out_high(&pin_lcd_cs);
 
     // Done initializing.
     pbdrv_init_busy_down();
 
-    // Regularly update the display with the user frame buffer, if changed.
-    etimer_set(&etimer, 40);
+    // Update the display with the user frame buffer, if changed.
     for (;;) {
-        PROCESS_WAIT_EVENT_UNTIL(ev == PROCESS_EVENT_TIMER && etimer_expired(&etimer));
-        if (pbdrv_display_user_frame_update_requested) {
-            pbdrv_display_user_frame_update_requested = false;
-            pbdrv_display_st7586s_encode_user_frame();
-            pbdrv_display_st7586s_write_data_begin(st7586s_send_buf, sizeof(st7586s_send_buf));
-            PROCESS_WAIT_EVENT_UNTIL(ev == PROCESS_EVENT_POLL && spi_status == SPI_STATUS_COMPLETE);
-            pbdrv_gpio_out_high(&pin_lcd_cs);
-        }
-        etimer_reset(&etimer);
+        PBIO_OS_AWAIT_UNTIL(state, pbdrv_display_user_frame_update_requested);
+        pbdrv_display_user_frame_update_requested = false;
+        pbdrv_display_st7586s_encode_user_frame();
+        pbdrv_display_st7586s_write_data_begin(st7586s_send_buf, sizeof(st7586s_send_buf));
+        PBIO_OS_AWAIT_UNTIL(state, spi_status == SPI_STATUS_COMPLETE);
+        pbdrv_gpio_out_high(&pin_lcd_cs);
     }
 
-    PROCESS_END();
+    PBIO_OS_ASYNC_END(PBIO_SUCCESS);
+}
+
+/**
+ * Initialize the display SPI driver.
+ *
+ * Pinmux and common EDMA handlers are already set up in platform.c.
+ */
+void pbdrv_display_init(void) {
+
+    // GPIO Mux. CS is in GPIO mode (manual control).
+    pbdrv_gpio_alt(&pin_spi1_mosi, SYSCFG_PINMUX5_PINMUX5_23_20_SPI1_SIMO0);
+    pbdrv_gpio_alt(&pin_spi1_clk, SYSCFG_PINMUX5_PINMUX5_11_8_SPI1_CLK);
+    pbdrv_gpio_input(&pin_lcd_a0);
+    pbdrv_gpio_input(&pin_lcd_cs);
+    pbdrv_gpio_out_high(&pin_lcd_reset);
+
+    // Waking up the SPI1 instance.
+    PSCModuleControl(SOC_PSC_1_REGS, HW_PSC_SPI1, PSC_POWERDOMAIN_ALWAYS_ON, PSC_MDCTL_NEXT_ENABLE);
+
+    // Reset.
+    SPIReset(SOC_SPI_1_REGS);
+    SPIOutOfReset(SOC_SPI_1_REGS);
+
+    // Mode.
+    uint32_t spipc0 = SPI_SPIPC0_SOMIFUN | SPI_SPIPC0_SIMOFUN | SPI_SPIPC0_CLKFUN | SPI_SPIPC0_ENAFUN;
```

Comment:
`SPI_SPIPC0_SOMIFUN` isn't actually wired up, so we could leave this out.

I don't think we are using `SPI_SPIPC0_ENAFUN` either.

Reply:
I did not change this code, I moved the function below `pbdrv_display_ev3_process_thread`. I change it to split it in `pbdrv_display_ev3_spi_init` and `pbdrv_display_init` so it’s clearer.